Arriba/Arriba.Web/parts/Search.jsx

```diff
@@ -53,60 +49,75 @@ export default React.createClass({
             }).cleaned);
         }
 
-        return Object.assign(this.getEmptyState(), {
-            allBasics: [],
+        this.state = {
             tables: [],
             page: 0,
-            hasMoreData: false,
             query: this.props.params.q || "",
             currentTable: table,
             currentTableSettings: {}, // {} denote no state, do not set to null.
             userSelectedTable: table,
             userSelectedId: this.props.params.open
-        });
-    },
-    componentDidMount: function () {
-        window.addEventListener("beforeunload", this); // For Mru
-        this.mru = new Mru();
+        };
+
+        this.events = {
+            "beforeunload": e => this.mru.push(),
+            "storage": e => { if (e.key.startsWith("table-")) this.getTableSettings() },
+        };
+    }
+    componentDidMount() {
+        super.componentDidMount();
         this.componentDidUpdate({}, {});
-    },
-    componentDidUpdate: function(prevProps, prevState) {
+    }
+    componentDidUpdate(prevProps, prevState) {
         const diffProps = Object.diff(prevProps, this.props);
-        const diff = Object.diff(prevState, this.state);
+        const diffState = Object.diff(prevState, this.state);
+
+        if (diffProps.hasAny("allBasics") || diffState.hasAny("debouncedQuery")) {
+            this.getCounts();
+        }
+
+        if (diffState.hasAny("userSelectedTable", "counts")) {
+            const currentTable = this.state.userSelectedTable || this.state.counts && this.state.counts.resultsPerTable[0].tableName;
+            this.setState({ currentTable: currentTable });
+        }
+
+        // Do not wipe userSelectedId if currentTable is going from `undefined` to defined.
+        // Scnario: On page load with open=something and table inferred from query.
+        if (diffState.hasAny("currentTable") && prevState.currentTable) {
+            this.setState({ userSelectedId: undefined });
+        }
+
+        if (diffProps.hasAny("allBasics") || diffState.hasAny("currentTable")) {
+            this.getTableSettings();
+        }
 
-        if (diffProps.hasAny("allBasics")) {
-            this.getAllCounts();
+        if (diffState.hasAny("debouncedQuery", "currentTableSettings", "page")) {
+            this.getListings();
         }
 
         // Watching currentTable as sometimes inferred from the query.
         // Not watching for query changes (to match old behavior).
-        if (diffProps.hasAny("allBasics") || diff.hasAny("currentTable", "userSelectedId")) {
+        if (diffProps.hasAny("allBasics") || diffState.hasAny("currentTable", "userSelectedId")) {

```

Comment:
Not watching for query changes here is tricky. Basically we should load details again when the query changes (to update highlighting) but we *currently* have a behavior where we change the userSelectedId to undefined when the query changes, so it doesn't currently matter because we'll close the opened item.

I would argue we should code this to re-get details if the query changes so that if we decided to leave open items open as the query changes, the highlighting would update. I would not expect this change to have a negative perf impact because the debouncedQuery and userSelectedId should be changing in the same setState.

Reply:
Thanks for bringing up the subtlety of updating the highlighting. I agree on the ideal behavior. I'm assuming it's not a pressing issue, is that correct?

Reply:
Yes, not pressing.